### PR TITLE
Move TCP connection away from Mac Network lib over to  swift-nio enabling

### PR DESCRIPTION
**Pull request by shial4**
Move TCP connection away from Mac Network lib over to  swift-nio enabling, by doing this we enable framework to work wherever Swift does. such as Linux, Windows, even server side. 

@stensoosaar 
I've tried to keep everything as you had with modifying the connection layer in most. 
I haven't had a chance to test it with the IB as IU will get my access back after 15th next month. I'm counting here on your help. 

I will leave comments below to address what I think is important.
Let me know what you think :)!

### CODE REVIEW

File: IBKit/IBKit/Client/IBClientFrameDecoder.swift [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    func decodeLast(buffer: inout ByteBuffer, seenEOF: Bool) throws -> InboundOut? {
+        let decoded = try self.decode(buffer: &buffer)
+        if buffer.readableBytes > 0 {
+            throw IBError.failedToRead("leftover bytes")
+        }
+        return decoded
+    }
+
+    private func readFrame(buffer: inout ByteBuffer) throws -> ByteBuffer? {
```
Comment (shial4) [inline IBKit/IBKit/Client/IBClientFrameDecoder.swift:61]: Not sure if I did implemented correctly frames decoding, but I left some comments here. Let me know how does it look.

This handler is mostly responsible for reading data and packaging them up into frames.

File: IBKit/IBKit/Client/IBMessageHandler.swift [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+//    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+//    SOFTWARE.
+//
+
+import Foundation
+import NIOCore
+
+class IBMessageHandler: ChannelInboundHandler {
```
Comment (shial4) [inline IBKit/IBKit/Client/IBMessageHandler.swift:27]: This is our massage handler, that simply converts frames in data and we pass them to our IBConnection

**Comment by stensoosaar** [comment]
To get us into same page i included your proposal as NIOClient etc into the dev branch. Hopefully didn't miss anything crucial. At the moment it seems to stuck at opening connection though and never reach connected state to write the greeting message.

**Comment by shial4** [comment]
Yea, I've fixed it already. Pull the recent changes. Although it disconnects now when I try to send any message. Like subscribe to real time data

**Comment by shial4** [comment]
@stensoosaar let me convert this PR into draft. Till I fix the connection so it is reliable.

File: IBKit/IBKit/Client/IBClientFrameDecoder.swift [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        guard buffer.readableBytes >= 4 else {
+            return nil
+        }
+
+        let lengthPrefix = buffer.getInteger(at: buffer.readerIndex, as: Int32.self)!
+        let frameLength = Int(lengthPrefix.littleEndian)
```
Comment (shial4) [inline IBKit/IBKit/Client/IBClientFrameDecoder.swift:65]: FYI:
I had error here, thought it is bigEndian, but it does appear to be littleEndian

File: IBKit/IBKit/Client/IBConnection.swift [hunk 4]
```diff
@@ -24,143 +24,185 @@
...
-		if let didStopCallback = self.didStopCallback {
-			self.didStopCallback = nil
-			didStopCallback(error)
-		}
-	}
+    enum State: Equatable {
+        case initializing
+        case connecting(String)
+        case connected
+        case connectedToAPI
```
Comment (shial4) [inline IBKit/IBKit/Client/IBConnection.swift:38]: also added, state indicating we are connected to the API

**Comment by stensoosaar** [comment]
Many thanks! Its working! One more thing though... Could you review the connection locking logic, as there seems to be outgoing message race between user requests and handshaking process...
```
connecting: 127.0.0.1:4002
connected

1714051808.505105.    ("API\0\0\0\0\tv142..175")
1714051808.5111098.   ("\0\0\0G20\00\00\0ETH\0CRYPTO\0\00.00\0\00.00\0PAXOS\0\0USD\0\0\00\0\01 min\060 S\00\0TRADES\01\01\0\0")
connected to API
1714051808.526440.   ("\0\0\0\n71\02\0999\0\0")

**Pull request merged**